danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt

```diff
@@ -0,0 +1,60 @@
+package systems.danger.kotlin
+
+// extensions over [Git] object
+
+/**
+ * Changed lines in this PR
+ */
+val Git.changedLines: PullRequestChangedLines
+    get() {
+        // TODO replace with more generic way to process commands
+        val utils = Utils()
+        val commandRawOutput = utils.exec("git diff --numstat $headSha $baseSha")
+        val additionDeletionPairs = commandRawOutput.lines()
+            .filter { it.isNotEmpty() }
+            .map { line ->
+                val parts = line.split("\\s+".toRegex())
+                parts[0].toInt() to parts[1].toInt()
+            }
+        val additions = additionDeletionPairs.fold(0) { acc, (addition, _) -> acc + addition }
+        val deletions = additionDeletionPairs.fold(0) { acc, (_, deletion) -> acc + deletion }
+        return PullRequestChangedLines(additions, deletions)
+    }
+
+/**
+ * Number of changed lines
+ */
+val Git.linesOfCode: Int
+    get() = additions + deletions
+
+/**
+ * Number of added lines
+ */
+val Git.additions: Int
+    get() = changedLines.additions
+
+/**
+ * Number of deleted lines
+ */
+val Git.deletions: Int
+    get() = changedLines.deletions
+
+/**
+ * Reference to a SHA of head commit of this PR
+ */
+val Git.headSha: String?
+    get() = commits.first().sha
+
+/**
+ * Reference to a SHA of base commit of this PR
+ */
+val Git.baseSha: String?
+    get() = "${commits.last().sha}^1"
```

Comment:
Why `^1`? I think your base is the actual `commits.last().sha`
If you use this command here you would skip the changes from `c22ea38` I think, am I missing something?

Reply:
Maybe I am missing something, but if I take this PR as an example, so here are three commits:
* 7370a5fa2c0720c3e38322b75603687aea0f1d6a
* 3e1bd1539b12d885f3943995bbfcc9c275bf11a2
* c22ea384100a297780be7b0499ba6687a407e0b6

if I compute diff like
```
git diff --numstat c22ea384100a297780be7b0499ba6687a407e0b6 7370a5fa2c0720c3e38322b75603687aea0f1d6a
```
the result is 
```
0       2       CHANGELOG.md
19      20      danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt
```
it will return changes that happened from the commit c22ea384100a297780be7b0499ba6687a407e0b6 to 7370a5fa2c0720c3e38322b75603687aea0f1d6a but it does not take changes in the 7370a5fa2c0720c3e38322b75603687aea0f1d6a into account. So if I take the changes from the parent of the first commit up until the last commit we know, in our case
```
git diff --numstat c22ea384100a297780be7b0499ba6687a407e0b6 7370a5fa2c0720c3e38322b75603687aea0f1d6a^1
```
it will return correctly
```
0       2       CHANGELOG.md
1       2       danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/Git.kt
0       60      danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/GitKtx.kt
3       3       danger-kotlin-library/src/main/kotlin/systems/danger/kotlin/Utils.kt
```
which if summed up matches the number [here](https://github.com/danger/kotlin/pull/102/commits)

Also if you would have only a single commit in the PR, eg. in our repo [here](https://github.com/AckeeCZ/ackee-gradle-plugin/pull/13/commits) the command
```
git diff --numstat 49ec200b25a1130cfb86933f41bbf79e40b997bf
```
produces
```
4       0       README.md
33      8       build-gradle-plugin/src/main/kotlin/cz/ackee/gradle/AckeePluginKotlin.kt
1       1       lib.properties
```
which does not match. But if I do this
```
git diff --numstat 49ec200b25a1130cfb86933f41bbf79e40b997bf 49ec200b25a1130cfb86933f41bbf79e40b997bf^1
```
it produces that
```
0       31      .github/workflows/bintray_upload.yml
4       12      gradle/bintray/bintray.gradle
```
which is correct. 

So maybe I am missing something or I am doing it ineffciently, but from my testing this method works. 

Basically I want to see all changes from the commit my branch is originating from up until the last commit in my PR. That's how I understand it

Reply:
Ok it makes sense, the last commit is then the actual first commit you made.
Now it makes sense, thank you :)